File: android/app/src/main/java/es/manabe/yomiyasu/core/services/SocketService.kt
package es.manabe.yomiyasu.core.services

import android.util.Log
import es.manabe.yomiyasu.app.ServerConfig
import io.socket.client.IO
import io.socket.client.Manager
import io.socket.client.Socket
import io.socket.engineio.client.transports.WebSocket
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import org.json.JSONObject
import java.net.URI
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class SocketService @Inject constructor() {

    private val _isConnected = MutableStateFlow(false)
    val isConnected: StateFlow<Boolean> = _isConnected.asStateFlow()

    private val _libraryUpdatedAt = MutableStateFlow<Long?>(null)
    val libraryUpdatedAt: StateFlow<Long?> = _libraryUpdatedAt.asStateFlow()

    private var manager: Manager? = null
    private var socket: Socket? = null

    var enabled: Boolean = true

    fun start() {
        if (!enabled) return
        if (manager != null) return

        try {
            val serverUrl = ServerConfig.serverUrl
            if (serverUrl == null) {
                Log.w(TAG, "Sin servidor configurado; websocket no iniciado")
                return
            }

            val options = IO.Options().apply {
                path = "/socket.io/"
                transports = arrayOf(WebSocket.NAME)
                reconnection = true
                reconnectionDelay = 2_000
            }

            val uri = URI.create(serverUrl.toString())
            val newManager = Manager(uri, options)
            manager = newManager

            val newSocket = newManager.socket("/ws")
            socket = newSocket

            newSocket.on(Socket.EVENT_CONNECT) {
                _isConnected.value = true
                Log.i(TAG, "Websocket conectado")
            }

            newSocket.on(Socket.EVENT_DISCONNECT) {
                _isConnected.value = false
                Log.i(TAG, "Websocket desconectado")
            }

            newSocket.on("notification") { args ->
                handleNotification(args)
            }

            newSocket.connect()
            Log.i(TAG, "Conectando al websocket $serverUrl")
        } catch (error: Exception) {
            Log.e(TAG, "No se ha podido iniciar el websocket", error)
        }
    }

    fun stop() {
        runCatching { socket?.disconnect() }
        runCatching { socket?.close() }
        socket = null
        manager = null
        _isConnected.value = false
    }

    private fun handleNotification(args: Array<Any>) {
        val payload = args.firstOrNull() ?: return

        val action = when (payload) {
            is JSONObject -> payload.optString("action", "")
            is Map<*, *> -> payload["action"] as? String ?: ""
            is String -> runCatching { JSONObject(payload).optString("action", "") }.getOrDefault("")
            else -> ""
        }

        if (action.isEmpty()) return

        Log.i(TAG, "Notificación recibida: $action")

        if (action == "LIBRARY_UPDATE") {
            markLibraryUpdated()
        }
    }

    /**
     * Provoca la recarga de las vistas que observan ``libraryUpdatedAt``
     * tras una mutación local (marcar leído, leer más tarde, pausar serie…).
     */
    fun markLibraryUpdated() {
        _libraryUpdatedAt.value = System.currentTimeMillis()
    }

    private companion object {
        const val TAG = "SocketService"
    }
}

File: ios/Yomiyasu/Core/Services/SocketService.swift
import Foundation
import Observation
import SocketIO
import os

@MainActor
@Observable
final class SocketService {
    private(set) var isConnected = false
    private(set) var libraryUpdatedAt: Date?

    private var url: URL
    private let logger = Logger(subsystem: "es.manabe.yomiyasu", category: "SocketService")

    private var manager: SocketManager?
    private var socket: SocketIOClient?

    init(url: URL = APIClient.unconfiguredBaseURL) {
        self.url = url
    }

    func setURL(_ url: URL) {
        self.url = url
    }

    func start() {
        guard manager == nil, url != APIClient.unconfiguredBaseURL else { return }

        logger.info("Conectando al websocket \(self.url.absoluteString, privacy: .public)")

        let manager = SocketManager(
            socketURL: url,
            config: [
                .compress,
                .forceWebsockets(true),
                .path("/socket.io/"),
                .reconnects(true),
                .reconnectWait(2),
            ]
        )
        self.manager = manager

        let socket = manager.socket(forNamespace: "/ws")
        self.socket = socket

        socket.on(clientEvent: .connect) { [weak self] _, _ in
            Task { @MainActor in
                self?.isConnected = true
                self?.logger.info("Websocket conectado")
            }
        }

        socket.on(clientEvent: .disconnect) { [weak self] _, _ in
            Task { @MainActor in
                self?.isConnected = false
                self?.logger.info("Websocket desconectado")
            }
        }

        socket.on("notification") { [weak self] data, _ in
            Task { @MainActor in
                self?.handleNotification(data)
            }
        }

        socket.connect()
    }

    func stop() {
        socket?.disconnect()
        manager?.disconnect()
        socket = nil
        manager = nil
        isConnected = false
    }

    func handleNotification(_ data: [Any]) {
        guard let payload = data.first as? [String: Any],
              let action = payload["action"] as? String else {
            return
        }

        logger.info("Notificación recibida: \(action, privacy: .public)")

        if action == "LIBRARY_UPDATE" {
            libraryUpdatedAt = .now
        }
    }

    /// Provoca la recarga de las vistas que observan ``libraryUpdatedAt``
    /// tras una mutación local (marcar leído, leer más tarde, pausar serie…).
    func markLibraryUpdated() {
        libraryUpdatedAt = .now
    }
}
